euphonic/spectra.py

```diff
@@ -669,10 +674,387 @@ def broaden(self: T, x_width,
+        """
+        metadata = copy.deepcopy(self.metadata)
+        metadata.pop('line_data', None)
+        metadata.update(self._tidy_metadata())
+        summed_s_data = ureg.Quantity(
+            np.sum(self._get_raw_spectrum_data(), axis=0),
+            units=self._get_internal_spectrum_data_unit()
+        ).to(self._get_spectrum_data_unit())
+        return self._item_type(
+            **self._get_bin_kwargs(),
+            **{self._spectrum_data_name(): summed_s_data},
+            x_tick_labels=copy.copy(self.x_tick_labels),
+            metadata=metadata
+        )
+
+    # Required methods
+    @classmethod
+    @abstractmethod
+    def from_spectra(cls, spectra: Sequence[Spectrum]) -> Self:
+        """Construct spectrum collection from a sequence of components"""
+        ...
+
+    # Mixin methods
+    def __len__(self):
+        return self._get_raw_spectrum_data().shape[0]
+
+    @overload
+    def __getitem__(self, item: int) -> Spectrum:
+        ...
+
+    @overload  # noqa: F811
+    def __getitem__(self, item: slice) -> Self:
+        ...
+
+    @overload  # noqa: F811
+    def __getitem__(self, item: Union[Sequence[int], np.ndarray]) -> Self:
+        ...
+
+    def __getitem__(
+            self, item: Union[Integral, slice, Sequence[Integral], np.ndarray]
+    ):  # noqa: F811
+        self._validate_item(item)
+        init_kwargs = {
+            self._spectrum_data_name(): self._get_spectrum_data()[item, :],
+            "x_tick_labels": self.x_tick_labels,
+            "metadata": self._get_item_metadata(item)
+                       } | self._get_bin_kwargs()
+
+        if isinstance(item, Integral):
+            return self._item_type(**init_kwargs)
+
+        return type(self)(**init_kwargs)
+
+    def _validate_item(self, item: Integral | slice | Sequence[Integral] | np.ndarray
+                       ) -> None:
+        """Raise Error if index has inappropriate typing/range"""
+        if isinstance(item, Integral):
+            return
+        if isinstance(item, slice):
+            if (item.stop is not None) and (item.stop >= len(self)):
+                raise IndexError(f'index "{item.stop}" out of range')
+            return
+
+        if not all([isinstance(i, Integral) for i in item]):
+            raise TypeError(
+                f'Index "{item}" should be an integer, slice '
+                f'or sequence of ints')
+
+    @overload
+    def _get_item_metadata(self, item: Integral) -> OneLineData:
+        """Get a single metadata item with no line_data"""
+
+    @overload
+    def _get_item_metadata(self, item: slice | Sequence[Integral] | np.ndarray
+                           ) -> Metadata:  # noqa: F811
+        """Get a metadata collection (may include line_data)"""
+
+    def _get_item_metadata(self, item):  # noqa: F811
+        """Produce appropriate metadata for __getitem__"""
+        metadata_lines = list(self.iter_metadata())
+
+        if isinstance(item, Integral):
```

Comment:
I initially decided against it because of the special case for `len(item) == 1`, which doesn't fit.

But on reflection that special case is a bad idea and not indicated properly in the type  hint anyway. Let's see how much harm is really done by removing it...

Reply:
Actually the special case was not so bad, it deals with an inconsistency in output of `itemgetter`. Instead we can use

```python
return _combine_metadata([metadata_lines[i] for i in item])
```
for all sequence cases.

I tried to use singledispatchmethod but it didn't go well:
- doesn't support Union types so Sequence[int] and np.ndarray have to be handled separately
- doesn't support Sequence[int] at all
- at one point I was getting weird abstract class errors that child classes were obliged to implement their own `singledispatchmethod` function.

As far as I can tell, then, the standard library `singledispatch` features are pretty useless 😞 
The Union types may have been fixed in a more recent Python version?

Reply:
Yeah, way to to union types is:
```
x.register(int)
x.register(float)
def _(...)
```
But it doesn't support subscripted types which is a pain.

Reply:
I tried it with bare `Sequence` but then I hit the "abstract method" weirdness.

Reply:
Sadly abandoning this approach for now. I wish it worked.

Reply:
Yeah. More important to get this merged!